providers/enchant_nuspell.cpp

```diff
@@ -0,0 +1,438 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <string>
+#include <vector>
+
+#include "enchant-provider.h"
+#include "unused-parameter.h"
+
+#include <nuspell/dictionary.hxx>
+#include <nuspell/finder.hxx>
+
+#include <glib.h>
+
+using namespace std;
+using namespace nuspell;
+
+/***************************************************************************/
+
+class NuspellChecker
+{
+public:
+	bool checkWord (const char *word, size_t len);
+	char **suggestWord (const char* const word, size_t len, size_t *out_n_suggs);
+
+	bool requestDictionary (const char * szLang);
+
+private:
+	Dictionary nuspell;
+};
+
+/***************************************************************************/
+
+bool
+NuspellChecker::checkWord(const char *utf8Word, size_t len)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+
+	return nuspell.spell(string(normalizedWord));
+}
+
+char**
+NuspellChecker::suggestWord(const char* const utf8Word, size_t len, size_t *nsug)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+	auto suggestions = vector<string>();
+	nuspell.suggest(string(normalizedWord), suggestions);
+	if (suggestions.empty())
+		return nullptr;
+	*nsug = suggestions.size();
+	char **sug = g_new0 (char *, *nsug + 1);
+	size_t i = 0;
+	for (auto& suggest : suggestions) {
+		char *word = g_new0(char, suggest.size() + 1);
+		strcpy(word, suggest.c_str());
+		sug[i] = word;
+		i++;
+	}
+	return sug;
+}
+
+static void
+s_buildDictionaryDirs (vector<string> & dirs)
+{
+	dirs.clear ();
+
+	/* 1. personal overrides for Enchant
+	 *    ~/.config/enchant/nuspell
+	 */
+	gchar * tmp;
+	char * config_dir = enchant_get_user_config_dir ();
+	tmp = g_build_filename (config_dir, "nuspell", nullptr);
+	dirs.push_back (tmp);
+	free (config_dir);
+	g_free(tmp);
+
+	/* Dynamically retieved from Nuspell dictionary finder:
+	 * 2. personal overrides for Hunspell
+	 *    ~/.local/share/hunspell
+	 * 3. system installed for Hunspell
+	 *    /usr/local/share/hunspell
+	 *    /usr/share/hunspell
+	 * 4. personal or system installations for
+	 *    Mozilla (such as Firefox and Thunderbird),
+	 *    LibreOffice and Apache OpenOffice
+	 */
+	auto nuspell_finder = nuspell::Finder::search_all_dirs_for_dicts();
+	for (auto& path : nuspell_finder.get_dir_paths()) {
+		if (path.compare(".") != 0)
+			dirs.push_back (path.c_str());
+	}
+
+	/* 5. system installations by Enchant
+	 *    /usr/local/share/enchant/nuspell
+	 *    /usr/share/enchant/nuspell
+	 */
+	char * enchant_prefix = enchant_get_prefix_dir();
+	if (enchant_prefix) {
+		tmp = g_build_filename(enchant_prefix, "share", "enchant", "nuspell", nullptr);
+		dirs.push_back (tmp);
+		g_free(enchant_prefix);
+		g_free(tmp);
+	}
+
+	/* Hunspell paths are used, therefore ENCHANT_NUSPELL_DICT_DIR is
+	 * irrelevant. Hence, the following paths are not to be considered:
```

Comment:
Should this simply say "Hence, the following paths are not considered"?

Reply:
When saying only 'are not considered' it might seems like a temporary issue/fix, but these paths should not be considered now and nor in the near to intermediate future.